src/libstd/collections/hash/map.rs

```diff
@@ -2362,9 +1844,9 @@ impl<'a, K: 'a + Debug, V: 'a + Debug> Debug for OccupiedEntry<'a, K, V> {
 /// [`Entry`]: enum.Entry.html
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct VacantEntry<'a, K: 'a, V: 'a> {
-    hash: SafeHash,
+    hash: u64,
     key: K,
-    elem: VacantEntryState<K, V, &'a mut RawTable<K, V>>,
+    table: &'a mut RawTable<(K, V)>,
```

Comment:
There are different constraints when searching for an element vs searching for an insertion slot. In particular, we ignore (skip over) deleted buckets in the former case while in the latter case we can insert on top of a deleted bucket.

Reply:
Apologies for the confusing comment. I think the `Entry` API allows two implementations here. On return:
1) `VacantEntry` already points to the actual bucket and `insert()` is trivial.
2) `VacantEntry` caches the key and hash and `insert()` does the actual insert/grow dance.

In either case the optimal way to implement _find_or_insert_ semantics needed by `entry() -> insert()` is:
- find probe, if found return it (`insert()` won't be called)
- insert probe, if bucket is deleted or if `growth_left > 0` return it (this probe costs very little because we are touching at most the same cachelines we touched in the previous probe - a deleted bucket can only make us return earlier)
- rehash and do another insert probe to find where to insert it (the cost of an additional probe here is insignificant - we already did N probes to rehash the container and copied N key, value tuples)